src/LegacyThumbnailMigrationHelper.php

```diff
@@ -0,0 +1,197 @@
+     * Perform migration
+     *
+     * @param AssetStore $store
+     * @return array Map of old to new moved paths
+     */
+    public function run(AssetStore $store)
+    {
+        // Set max time and memory limit
+        Environment::increaseTimeLimitTo();
+        Environment::increaseMemoryLimitTo();
+
+        // Loop over all folders
+        $allMoved = [];
+        $originalState = null;
+        if (class_exists(Versioned::class)) {
+            $originalState = Versioned::get_reading_mode();
+            Versioned::set_stage(Versioned::DRAFT);
+        }
+
+        // Migrate root folder (not returned from query)
+        $moved = $this->migrateFolder($store, new Folder());
+        if ($moved) {
+            $allMoved = array_merge($allMoved, $moved);
+        }
+
+        // Migrate all nested folders
+        $folders = $this->getFolderQuery();
+        foreach ($folders->dataQuery()->execute() as $folderData) {
+            // More memory efficient way to looping through large sets
+            /** @var Folder $folder */
+            $folder = $folders->createDataObject($folderData);
+            $moved = $this->migrateFolder($store, $folder);
+            if ($moved) {
+                $allMoved = array_merge($allMoved, $moved);
+            }
+        }
+        if (class_exists(Versioned::class)) {
+            Versioned::set_reading_mode($originalState);
+        }
+        return $allMoved;
+    }
+
+    /**
+     * @param LoggerInterface $logger
+     */
+    public function setLogger(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+
+        return $this;
+    }
+
+    /**
+     * Migrate a folder
+     *
+     * @param AssetStore $store
+     * @param Folder $folder
+     * @param string $legacyFilename
+     * @return array Map of old to new file paths (relative to store root)
+     */
+    protected function migrateFolder(AssetStore $store, Folder $folder)
+    {
+        $moved = [];
+
+        // Get normalised store relative path
+        $folderPath = preg_replace('#^/#', '', $folder->getFilename());
+        $resampledFolderPath = $folderPath . '_resampled';
+
+        // Legacy thumbnails couldn't have been stored in a protected filesystem
+        /** @var \League\Flysystem\Filesystem $filesystem */
+        $filesystem = $store->getPublicFilesystem();
+
+        /** @var FileResolutionStrategy $strategy */
+        $strategy = $store->getPublicResolutionStrategy();
+
+        if (!$filesystem->has($resampledFolderPath)) {
+            return $moved;
+        }
+
+        // Recurse through folder
+        foreach ($filesystem->listContents($resampledFolderPath, true) as $fileInfo) {
+            if ($fileInfo['type'] !== 'file') {
+                continue;
+            }
+
+            $oldResampledPath = $fileInfo['path'];
+
+            // Get "variant folders" inside _resampled folder.
+            $variantFolders = preg_replace('#.*_resampled/(.*)#', '$1', $fileInfo['dirname']);
```

Comment:
Dunno, I'd need to explode the resulting variant on underscores then, instead of just dealing with the directories. So double conversion and more moving parts in order to remove hardcoding `_resampled` in one place - feels low value.

Reply:
`LegacyFileIDHelper` already swaps the `/` for underscores.

Reply:
Yeah that's my point: I need the variants individually, not in `variant1/variant2` or `variant1_variant2` format. So I need to mess with strings anyway, even if I'm using `LegacyFileIDHelper`